apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/internal/AJResponseWriter.kt

```diff
@@ -0,0 +1,225 @@
+package com.apollographql.apollo.cache.normalized.internal
+
+import com.apollographql.apollo.api.BigDecimal
+import com.apollographql.apollo.api.CustomScalar
+import com.apollographql.apollo.api.CustomScalarAdapters
+import com.apollographql.apollo.api.Operation
+import com.apollographql.apollo.api.Response
+import com.apollographql.apollo.api.ResponseField
+import com.apollographql.apollo.api.internal.ResponseWriter
+import com.apollographql.apollo.api.internal.Utils.shouldSkip
+import com.apollographql.apollo.cache.normalized.CacheKey
+import com.apollographql.apollo.cache.normalized.CacheKeyResolver
+import com.apollographql.apollo.cache.normalized.Record
+
+/**
+ *
+ */
+class AJResponseWriter(
+    private val operationVariables: Operation.Variables,
+    private val customScalarAdapters: CustomScalarAdapters,
+) : ResponseWriter, ResponseWriter.ListItemWriter {
+
+  private var rootObject: AJObject? = null
+  fun root(): AJObject {
+    if (rootObject == null) {
+      /**
+       * close the root object  if needed
+       */
+      check(objectStack.size == 1)
+      rootObject = AJObject(fields = objectStack.removeLast())
+    }
```

Comment:
not sure but is it possible to use `lazy` intializer here?

Reply:
Yup, way more idiomatic